pkg/project/resource/resource_project_user_test.go

```diff
@@ -354,24 +356,112 @@ func TestAccProjectMember_missing_user_ignored(t *testing.T) {
 			use_project_user_resource = true
 		}
 
+		{{ if .create_user }}
+
+		resource "artifactory_managed_user" "{{ .username }}" {
+			name     = "{{ .username }}"
+			email    = "{{ .email }}"
+			password = "Password1!"
+			admin    = false
+		}
+
+		{{ end }}
+
 		resource "project_user" "{{ .username }}" {
 			project_key = project.{{ .project_name }}.key
 			name = "{{ .username }}"
 			roles = {{ .roles }}
 			ignore_missing_user = true
+			{{ if .create_user }}
+			depends_on = [artifactory_managed_user.{{ .username }}]
+			{{ end }}
 		}
 	`
 
 	config := util.ExecuteTemplate("TestAccProjectUser", template, params)
+
+	updateParams := map[string]interface{}{
+		"project_name": params["project_name"],
+		"project_key":  params["project_key"],
+		"username":     params["username"],
+		"email":        params["email"],
+		"roles":        params["roles"],
+		"create_user":  true,
+	}
+
+	configUpdated := util.ExecuteTemplate("TestAccProjectUser", template, updateParams)
+
 	resource.Test(t, resource.TestCase{
 		PreCheck: func() { acctest.PreCheck(t) },
 		CheckDestroy: acctest.VerifyDeleted(resourceName, func(id string, request *resty.Request) (*resty.Response, error) {
 			return verifyProjectUser(username, projectKey, request)
 		}),
 		ProtoV6ProviderFactories: acctest.ProtoV6ProviderFactories,
 		Steps: []resource.TestStep{
+			// attempt create, will not work
 			{
 				Config: config,
+				ConfigPlanChecks: resource.ConfigPlanChecks{
+					PostApplyPostRefresh: []plancheck.PlanCheck{
+						// expect create of project user
+						plancheck.ExpectResourceAction(resourceName, plancheck.ResourceActionCreate),
+					},
+				},
+				// expect user to be added to state
+				Check: resource.ComposeTestCheckFunc(
```

Comment:
> 1. Add ExpectNonEmtpyPlan: true
Yes, this looks promising according to: https://github.com/hashicorp/terraform-plugin-testing/blob/4a889266a95c9a38734a7c9f6e079b641a8fae53/helper/resource/testing_new_config.go#L358-L376

> 2. Remove Check: ... since there'd be no state being saved by the resource.

TF will still write the project_user to state in this variant of the fix. My initial attempt to not write a state entry failed. TF will report an error if a TF provider does not create a state entry for a resource. So its either "fail" or "create state entry". 

The key change in this PR is that on a `read` we remove the item from the state if it is missing (causing a create to be detected on plan). Hence the assertions that the state has values as I've seen it on manual tests.

Reply:
This is the test `Steps` that pass for me:

```go
Steps: []resource.TestStep{
	// attempt create, will not work
	{
		Config: config,
		ConfigPlanChecks: resource.ConfigPlanChecks{
			PostApplyPostRefresh: []plancheck.PlanCheck{
				// expect create of project user
				plancheck.ExpectResourceAction(resourceName, plancheck.ResourceActionCreate),
			},
		},
		ExpectNonEmptyPlan: true,
		// expect user to be added to state
		Check: resource.ComposeTestCheckFunc(
			resource.TestCheckResourceAttr(resourceName, "project_key", fmt.Sprintf("%s", params["project_key"])),
			resource.TestCheckResourceAttr(resourceName, "name", username),
			resource.TestCheckResourceAttr(resourceName, "ignore_missing_user", "true"),
			resource.TestCheckResourceAttr(resourceName, "roles.#", "2"),
			resource.TestCheckResourceAttr(resourceName, "roles.0", "Developer"),
			resource.TestCheckResourceAttr(resourceName, "roles.1", "Project Admin"),
		)},
	// re-attempt create, will still not work
	{
		Config: config,
		ConfigPlanChecks: resource.ConfigPlanChecks{
			PostApplyPostRefresh: []plancheck.PlanCheck{
				// again expect create of project user (refresh will mark project user as missing)
				plancheck.ExpectResourceAction(resourceName, plancheck.ResourceActionCreate),
			},
		},
		ExpectNonEmptyPlan: true,
		// expect user to be in the state
		Check: resource.ComposeTestCheckFunc(
			resource.TestCheckResourceAttr(resourceName, "project_key", fmt.Sprintf("%s", params["project_key"])),
			resource.TestCheckResourceAttr(resourceName, "name", username),
			resource.TestCheckResourceAttr(resourceName, "ignore_missing_user", "true"),
			resource.TestCheckResourceAttr(resourceName, "roles.#", "2"),
			resource.TestCheckResourceAttr(resourceName, "roles.0", "Developer"),
			resource.TestCheckResourceAttr(resourceName, "roles.1", "Project Admin"),
		)},
	// re-attempt create with user being added, will work
	{
		Config: configUpdated,
		ConfigPlanChecks: resource.ConfigPlanChecks{
			PreApply: []plancheck.PlanCheck{
				// again expect create of project user (refresh will mark project user as missing)
				plancheck.ExpectResourceAction(resourceName, plancheck.ResourceActionCreate),
			},
			PostApplyPostRefresh: []plancheck.PlanCheck{
				// again expect create of project user (refresh will mark project user as missing)
				plancheck.ExpectResourceAction(resourceName, plancheck.ResourceActionNoop),
			},
		},
		// expect user to be in the state
		Check: resource.ComposeTestCheckFunc(
			resource.TestCheckResourceAttr(resourceName, "project_key", fmt.Sprintf("%s", params["project_key"])),
			resource.TestCheckResourceAttr(resourceName, "name", username),
			resource.TestCheckResourceAttr(resourceName, "ignore_missing_user", "true"),
			resource.TestCheckResourceAttr(resourceName, "roles.#", "2"),
			resource.TestCheckResourceAttr(resourceName, "roles.0", "Developer"),
			resource.TestCheckResourceAttr(resourceName, "roles.1", "Project Admin"),
		)},
	// now user is there, no action should be performed
	{
		Config: configUpdated,
		ConfigPlanChecks: resource.ConfigPlanChecks{
			PostApplyPostRefresh: []plancheck.PlanCheck{
				// again expect create of project user (refresh will mark project user as missing)
				plancheck.ExpectResourceAction(resourceName, plancheck.ResourceActionNoop),
			},
		},
		// expect user to be in the state
		Check: resource.ComposeTestCheckFunc(
			resource.TestCheckResourceAttr(resourceName, "project_key", fmt.Sprintf("%s", params["project_key"])),
			resource.TestCheckResourceAttr(resourceName, "name", username),
			resource.TestCheckResourceAttr(resourceName, "ignore_missing_user", "true"),
			resource.TestCheckResourceAttr(resourceName, "roles.#", "2"),
			resource.TestCheckResourceAttr(resourceName, "roles.0", "Developer"),
			resource.TestCheckResourceAttr(resourceName, "roles.1", "Project Admin"),
		)},
	},
})
```

Reply:
I also need to add this to `TestCase` for step 3 to work:

```go
ExternalProviders: map[string]resource.ExternalProvider{
	"artifactory": {
		Source: "jfrog/artifactory",
	},
},
```

Reply:
@ alexhung Highly appreciate your help here due to my lack of a test environment. I updated the PR with your proposed changes. The changes make sense and still fit my expectations on how things should work. I'm still fairly new to developing TF providers, a lot of unknowns in the testing framework 😁